storm/src/main/storm/mesos/schedulers/DefaultScheduler.java

```diff
@@ -0,0 +1,249 @@
+   * to a specific topology and store the state in "mesosWorkerSlotMap". This way, when Storm later calls schedule, we can just
+   * look up the "mesosWorkerSlotMap" for a list of available slots for the particular topology.
+   * 3. Given MesosWorkerSlot extends WorkerSlot, we shouldn't have to really create a "mesosWorkerSlotMap". Instead, in the schedule
+   * method, we could have just upcasted the "WorkerSlot" to "MesosWorkerSlot". But this is not currently possible because storm
+   * passes a recreated version of WorkerSlot to schedule method instead of passing the WorkerSlot returned by this method as is.
+    */
+  @Override
+  public List<WorkerSlot> allSlotsAvailableForScheduling(RotatingMap<Protos.OfferID, Protos.Offer> offers,
+                                                         Collection<SupervisorDetails> existingSupervisors,
+                                                         Topologies topologies, Set<String> topologiesMissingAssignments) {
+    if (topologiesMissingAssignments.isEmpty()) {
+      log.info("Declining all offers that are currently buffered because no topologies need assignments");
+      offers.clear();
+      return new ArrayList<>();
+    }
+
+    log.info("Topologies that need assignments: " + topologiesMissingAssignments.toString());
+
+    // Decline those offers that cannot be used for any of the topologies that need assignments.
+    for (Protos.Offer offer : offers.newestValues()) {
+      boolean isOfferUseful = false;
+      for (String currentTopology : topologiesMissingAssignments) {
+        boolean supervisorExists = SchedulerUtils.supervisorExists(offer.getHostname(), existingSupervisors, currentTopology);
+        TopologyDetails topologyDetails = topologies.getById(currentTopology);
+        if (SchedulerUtils.isFit(mesosStormConf, offer, topologyDetails, supervisorExists)) {
+          isOfferUseful = true;
+          break;
+        }
+      }
+      if (!isOfferUseful) {
+        log.info("Declining Offer " + offer.getId().getValue() + " because it does not fit any of the topologies that need assignments");
+        offers.clearKey(offer.getId());
+      }
+    }
+
+    List<WorkerSlot> allSlots = new ArrayList<>();
+
+    Map<String, List<OfferResources>> offerResourcesListPerNode = SchedulerUtils.getOfferResourcesListPerNode(offers);
+
+    for (String currentTopology : topologiesMissingAssignments) {
+      TopologyDetails topologyDetails = topologies.getById(currentTopology);
+      int slotsNeeded = topologyDetails.getNumWorkers();
+
+      double requestedWorkerCpu = MesosCommon.topologyWorkerCpu(mesosStormConf, topologyDetails);
+      double requestedWorkerMem = MesosCommon.topologyWorkerMem(mesosStormConf, topologyDetails);
+
+      log.info("Trying to find " + slotsNeeded + " slots for " + topologyDetails.getId());
+      if (slotsNeeded == 0) {
+        continue;
+      }
+
+      Map<String, Boolean> supervisorExistsPerNodeMap = new HashMap<>();
```

Comment:
This code has already been totally rewritten in our internal paired-on change.... I wouldn't spend *too* much time on reviewing this part of it, since you're gonna be rewriting/pushing our change for it very soon. ;-)